File: ClassB/09/homework2/task2.rb
require 'csv'
require 'matrix'

matrix1 = CSV.read(ARGV[0],converters: :numeric)
matrix2 = CSV.read(ARGV[1],converters: :numeric)

if matrix1.length != matrix2.length || matrix1[0].length != matrix2[0].length
	p 'undefined'
	return
end

rows = matrix1.length
cols = matrix1[0].length
if matrix1.length != matrix1[0].length
	p true
	return
end
result = Array.new(matrix1.length, 0){Array.new(matrix1[0].length, 0)}

for i  in 0..matrix1.length - 1
	for j in 0..matrix1[0].length - 1
		for k in 0..matrix1.length - 1
			#print "matrix1: ",matrix1[i][k],"  matrix2:",matrix2[k][j],"\n"
			result[i][j] += matrix1[i][k].to_i * matrix2[k][j].to_i
		end
	end
end

for i  in 0..matrix1.length - 1
	for j in 0..matrix1[0].length - 1
		if result[i][j] != (-1 * result[j][i])
			p true
			return
		end
	end
end

p false

File: classB/20/homework2/task2.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <fstream>

using namespace std;

int matrix_multiply(string, string);
int **read_matrix(string, int *, int *);

int main(int argc, char **argv) {
	int result;
	if(argc == 1) {
		result = matrix_multiply("matrix1.csv", "matrix2.csv");
	}
	else {
		result = matrix_multiply(argv[1], argv[2]);
	}
	if(result < 0) {
		printf("undefined\n");
	}
	else if(!result) {
		printf("true\n");
	}
	else {
		printf("false\n");
	}
	return 0;
}

int **read_matrix(string m, int *row, int *col) {
	ifstream fs(m);
	string line;

	int size_row = 100;
	int size_col = 100;
	int **matrix = 0;
	matrix = new int*[size_row];
	int i = 0;
	int j = 0;
	while(getline(fs, line)) {
		matrix[i] = new int[size_col];
		j = 0;
		istringstream in(line);
		while(!in.fail()) {
			in >> matrix[i][j];
			in.ignore();
			if(in.good()) {
				j++;
			}
		}
		if(i == 0) {
			*col = j + 1;
		}
		if(j != *col - 1) {
			*col = -1;
		}
		i++;
	}
	*row = i;
	fs.close();
	return matrix;
}

int matrix_multiply(string m1, string m2) {
	int row1 = 0, col1 = 0, row2 = 0, col2 = 0;
	int **matrix1 = read_matrix(m1, &row1, &col1);
	int **matrix2 = read_matrix(m2, &row2, &col2);

	int **multipy = new int[row1][col2];
	if(col1 == row2) {
		for(int i = 0; i < row1; i++) {
			for(int j2 = 0; j2 < col2; j2++) {
				for(int j = 0; j < col1; j++) {
					multipy[s][t] += (matrix1[i][j] * matrix2[j][j2]);
				}
			}
		}
	}
	else {
		return -1;
	}
	if(row1 == col1){
		for(int i = 0; i < row1; i++) {
			for(int j = 0; j < col1; j++) {
				if(multipy[i][j] != -multipy[j][i]) {
					return 1;
				}
			}
		}
		return 0;
	}
	else {
		return 1;
	}
}